firmware/Core/Src/telecommands/telecommand_args_helpers.c

```diff
@@ -13,16 +13,29 @@
 /// @param str Input string, starting with an integer
 /// @param str_len Max length of the input string
 /// @param result Pointer to the result
-/// @return 0 if successful, 1 if the string is empty, 2 if the string does not start with an integer
+/// @return 0 if successful, 1 if the string is empty, 2 if the input string doesn't match str_len, 3 for if the string is too long,
+///         4 if the result size doesn't match str_len (includes non-integer characters)
 uint8_t TCMD_ascii_to_uint64(const char *str, uint32_t str_len, uint64_t *result) {
-    // FIXME: return error if the string is too long/number is too large
-    // FIXME: return error if the number doesn't occupy the whole string (e.g., "123abc" with str_len=6 should error)
-    // TODO: write unit tests for this function
     // TODO: consider removing the str_len parameter and using strlen(str) instead (requires refactor in caller)
-    if (str_len == 0) {
+
+    // Error: the input string is empty
+    if (str_len == 0 || strlen(str) == 0) {
+        *result = 0;
         return 1;
     }
 
+    // Error: the input string doesn't match size of str_len
+    if(strlen(str) != str_len) {
```

Comment:
The way this function is setup to work (and is used) is actually such that you can pass it a pointer to a point within a string (not necessarily at the start), and only allow it to operate on a given length of the substring by setting the `str_len` to less than `strlen(str)`. This is a technique borrowed from Rust ("slicing" a string), which is a little uncommon in C.

This check, as-is, will cause this function to not work at all, I'm pretty sure. That's what my TODO comment was intended to indicate with the "requires refactor in caller" comment.